### PR TITLE
marius top screen stuff

### PR DESCRIPTION
**Pull request by marius851000**
Allow to display 16bpp images on top screen, and draw stuff in realtime following instruction.
Stuff I made for last hackjam but went unused (ported to US), only lightly tested (but the hackjam stuff was well tested).

Dunno if I’ll make use of it.

### CODE REVIEW

**Comment by Adex-8x** [comment]
Thank you so much! I'll check everything out soon, though first impressions:
- I assume the Top Screen has to be set into mode 0 by `back2_SetMode(0);`, which is alright, but might want to be noted somewhere.
- Your special processes do work on hardware/MelonDS, but after repeated use alongside `message_Menu(100);`, the game does seem to crash due to what I assume to be some memory-related issue? This might actually be a fault of how I handled the menu management, but I'll have to test further.
For reference, the script I tested with is:
```
def 0 {
    back2_SetMode(0);
    screen_FadeIn(1, 0);
    screen2_FadeIn(1, 0);
    ProcessSpecial(110, 1, 0);
    ProcessSpecial(PROCESS_SPECIAL_KEY_WAIT_PRESS, 0, 0);
    ProcessSpecial(111, 0, 0);
    ProcessSpecial(PROCESS_SPECIAL_KEY_WAIT_PRESS, 0, 0);
    ProcessSpecial(112, 1, 0);
    ProcessSpecial(PROCESS_SPECIAL_KEY_WAIT_PRESS, 0, 0);
    ProcessSpecial(111, 0, 0);
    ProcessSpecial(PROCESS_SPECIAL_KEY_WAIT_PRESS, 0, 0);
    end;
}
```
After this script runs twice, attempting to select an option from `message_Menu(100);` crashes the game.

File: src/drawing.c [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+void init_drawing(const char* filepath) {
+    DRAWING_STATE.x = 256/2;
+    DRAWING_STATE.y = 192/2;
+    DRAWING_STATE.is_pen_down = 0;
+    DRAWING_STATE.progress_per_frame = 5;
+    DRAWING_STATE.buffer_pos = 0;
+    DRAWING_STATE.file_closed = 0;
+    DRAWING_STATE.color_base = 0x8000;
+    DataTransferInit();
+    FileOpen(&DRAWING_STATE.file, filepath);
```
Comment (Adex-8x) [inline src/drawing.c:19]: Need to call [`FileInit`](https://github.com/UsernameFodder/pmdsky-debug/blob/4f361d8590e4ba28306670a20046e1c6ba38fa59/symbols/arm9/libs.yml#L2098-L2108) or [`FileInitVeneer`](https://github.com/UsernameFodder/pmdsky-debug/blob/4f361d8590e4ba28306670a20046e1c6ba38fa59/symbols/arm9.yml#L932-L944) prior to file opening.

File: src/display_image.c [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// under the MIT license, by marius851000
+
+#include "display_image.h"
+
+// require the top screen to be configured to LDLC mode, then switched back to BG mode for engine 2
+void displayImageOnConfiguredTopScreen(const char * filepath) {
+    struct file_stream file;
+    DataTransferInit();
+    FileOpen(&file, filepath);
```
Comment (Adex-8x) [inline src/display_image.c:10]: Need to call [`FileInit`](https://github.com/UsernameFodder/pmdsky-debug/blob/4f361d8590e4ba28306670a20046e1c6ba38fa59/symbols/arm9/libs.yml#L2098-L2108) or [`FileInitVeneer`](https://github.com/UsernameFodder/pmdsky-debug/blob/4f361d8590e4ba28306670a20046e1c6ba38fa59/symbols/arm9.yml#L932-L944) prior to file opening.

File: src/special_processes.c [hunk 3]
```diff
@@ -101,6 +104,27 @@ bool CustomScriptSpecialProcessCall(undefined4* unknown, uint32_t special_proces
...
+    // The marius stuff
+
+    // Display an image on top screen, initializing it if needed
+    // The script should make the top screen background does not change
+    case 110:
+      sprintf(temppath, "CUSTOM/SCREEN/%04d.raw", arg1);
+      COT_LOGFMT(COT_LOG_CAT_SPECIAL_PROCESS, "loading top screen raw %s", temppath);
+      displayImageOnTopScreen(temppath);
+      return true;
+    // Return top screen to what it was before it was set by 110 or 111
```
Comment (Adex-8x) [inline src/special_processes.c]: Do you mean "set by 110 or 112"?

**Review by Adex-8x — CHANGES_REQUESTED** [review]
Found the source of the crash, it's nothing tricky to fix yay

**Comment by marius851000** [comment]
Thanks a lot. Never met that crash situation before, but I still wonder how I could have left such an uninitialized variable error happen (that’s a joke. I totally knew how that happened, and why Rust is a better language than C, but that it’s not the most appropriate in this context).

**Review by Adex-8x — APPROVED** [review]
Thanks again! These are really cool features, hype

**Comment by marius851000** [comment]
Next, you’ll have a touchscreen trail following the cursor using the 3d engine. (also thinking about just throwing my custom textbox border drawing routine in the mix. It’s cool all you can do with the 3D engine)

**Pull request merged**